### PR TITLE
Set image width/height attributes in HTML

### PR DESCRIPTION
**Pull request by thewilkybarkid**
# ↪️ Pull Request

This sets the `width` and `height` attributes on HTML `` elements to help layout in browsers (refs https://github.com/parcel-bundler/parcel/issues/3738#issuecomment-549504886).

I had been using [posthtml-img-autosize](https://github.com/posthtml/posthtml-img-autosize) (with the `processEmptySize` option), but that's incompatible with things like the image transformer (since it runs before transformations).

## 💻 Examples

See the tests.

## 🚨 Test instructions

Tests included.

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by thewilkybarkid** [comment]
The potential layout problem is if someone has something like `img { width: 100% }` without `height: auto` (since the image will stretch). This is combination 3 on https://github.com/WICG/intrinsicsize-attribute/issues/16#issuecomment-500917583.

Or are using the width/height to control the size (refs https://github.com/parcel-bundler/parcel/pull/6783#discussion_r697663129 and https://github.com/parcel-bundler/parcel/issues/3738#issue-516907103).

Feels like the benefits far outweigh this edge case though.

File: packages/core/integration-tests/test/integration/html-image-size/index.html [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<!doctype html>
+<html>
+<body>
+<img id="image1" src="100x100.png">
+<img id="image2" src="100x100.png?width=150">
+<img id="image3" src="100x100.png?height=150">
+<img id="image4" src="100x100.png?width=200&height=150">
+<img id="image5" src="200x200.png" width="100">
+<img id="image6" src="200x200.png" height="100">
```
Comment (thewilkybarkid) [inline packages/core/integration-tests/test/integration/html-image-size/index.html:9]: As it stands, these are both fixed to be the real size of the image.

File: flow-libs/image-size.js.flow [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// @flow
+
+declare module 'image-size' {
+  declare module.exports: (input: Buffer) => { width: number, height: number, ...};
```
Comment (mischnic) [inline flow-libs/image-size.js.flow]: Apparently, `width` and `height` can also be undefined: https://github.com/image-size/image-size/blob/a38b56fe7303898ce6811adb213e518a57593d10/lib/types/interface.ts#L1-L10

Comment (thewilkybarkid) [reply flow-libs/image-size.js.flow]: I don't think those types match the implementation (I believe there should be union types of the different combinations), but better to match it.

**Comment by thewilkybarkid** [comment]
~I'm planning on experimenting with injecting a little CSS to avoid misshapen images (which is hopefully tree-shaken away in most cases).~ (Too much for now.)

File: packages/packagers/html/src/HTMLPackager.js [hunk 3]
```diff
@@ -112,6 +114,51 @@ async function getAssetContent(
...
+
+    const image = bundleGraph.getResolvedAsset(dependency);
+
+    if (!image) {
+      continue;
+    }
+
+    if (
+      typeof image.meta.width !== 'number' &&
+      typeof image.meta.height !== 'number'
```
Comment (devongovett) [inline packages/packagers/html/src/HTMLPackager.js]: Are these supposed to be set somewhere (e.g. the image transformer)?

File: packages/packagers/html/src/HTMLPackager.js [hunk 4]
```diff
@@ -112,6 +114,51 @@ async function getAssetContent(
...
+    ) {
+      const size = imageSize(await image.getBuffer());
+      image.meta.width = size.width;
+      image.meta.height = size.height;
```
Comment (devongovett) [inline packages/packagers/html/src/HTMLPackager.js]: I'm not sure we should mutate assets in packagers. I'm actually surprised this works... 😬

File: packages/packagers/html/src/HTMLPackager.js [hunk 5]
```diff
@@ -112,6 +114,51 @@ async function getAssetContent(
...
+    }
+
+    node.attrs.width = image.meta.width;
+    node.attrs.height = image.meta.height;
```
Comment (devongovett) [inline packages/packagers/html/src/HTMLPackager.js:142]: I think we should only do this if a `width` or `height` attribute is not already set on the element?

**Comment by mischnic** [comment]
We never want to make Parcel "break" your code after bundling, so if it worked correctly before bundling, it should still behave the same afterwards. That is unfortunately not always the case here:

https://codesandbox.io/s/serene-ives-legx7?file=/index.html

So I'm torn between automatically adding them to avoid layout shift and not randomly breaking the projects of unknowing users (especially because this is hard to debug if you don't know what you're looking for. In this case, you need to set the other "free" dimension to `auto` instead of `initial` to retain the aspect ratio).

File: packages/packagers/html/src/HTMLPackager.js (hunk 3 above)
Comment (thewilkybarkid) [reply packages/packagers/html/src/HTMLPackager.js]: Ideally, yes.

File: packages/packagers/html/src/HTMLPackager.js (hunk 5 above)
Comment (thewilkybarkid) [reply packages/packagers/html/src/HTMLPackager.js:142]: > I think we should only do this if a `width` or `height` attribute is not already set on the element?

Probably. I left this in for the moment (https://github.com/parcel-bundler/parcel/pull/6783#discussion_r697663129) for feedback (primarily about the CSS issue), but in the longer term it would be better for these to be "fixed" earlier in the pipeline (refs https://github.com/parcel-bundler/parcel/issues/3738#issue-516907103).

File: packages/packagers/html/src/HTMLPackager.js (hunk 4 above)
Comment (thewilkybarkid) [reply packages/packagers/html/src/HTMLPackager.js]: It could be that the image transformer sets the sizes (even if it's not actually transforming the image), and this sets them if present.

**Comment by thewilkybarkid** [comment]
> So I'm torn between automatically adding them to avoid layout shift and not randomly breaking the projects of unknowing users (especially because this is hard to debug if you don't know what you're looking for. In this case, you need to set the other "free" dimension to `auto` instead of `initial` to retain the aspect ratio).

I do wonder about this too; the packager doesn't feel the right place to do it but I'm not sure where is (the HTML transformer's too early?).

I did wonder about injecting some CSS (https://github.com/parcel-bundler/parcel/pull/6783#issuecomment-913039779), but that also could be thorny (and I'm not sure how and _where_ it should happen: inline and before any CSS in all HTML files that contain images?).

**Comment by thewilkybarkid** [comment]
If it helps, this could be split in two: the first part that adds the width/height as metadata on image assets, then a follow-up that uses it to set the width/height in HTML. Since the latter has potential CSS problems, I could release this as a separate plugin. (The former could also be a separate plugin, but doesn't seem unreasonable to make that information generally available for image assets?)

> the packager doesn't feel the right place to do it but I'm not sure where is (the HTML transformer's too early?).

Looking at the plugin system again, would it be an optimizer? (It's not 'optimization' as such, but needs to be a late-running transformation.)

**Comment by thewilkybarkid** [comment]
> Looking at the plugin system again, would it be an optimizer? (It's not 'optimization' as such, but needs to be a late-running transformation.)

I had a go at writing an optimizer and managed to get it to work, it just seems a bit tricky to find the asset from the `src` (which is now the final path, e.g. `/some-image.ab12345.jpg`).

First attempt looks like:

```js
tree.match({tag: 'img', attrs: {src: true}}, node => {
  let image;

  for (const dependency of dependencies) {
    const asset = bundleGraph.getResolvedAsset(dependency, bundle)

    if (!asset) {
      continue;
    }

    const bundles = bundleGraph.getBundlesWithAsset(asset);

    if (node.attrs['src'] !== bundles[0].name) {
      continue;
    }

    image = asset;
    break;
  }

  // use imageSize on image and update attributes on node

  return node;
});
```

which works, but is rather ugly. Is there a simple way that I've missed?

Does either of these approaches make sense @devongovett @mischnic? I've happy to update this PR or close it and release something standalone.

File: packages/core/integration-tests/test/integration/html-image-size/index.html (hunk 1 above)
Comment (thewilkybarkid) [reply packages/core/integration-tests/test/integration/html-image-size/index.html:9]: Would be changed by https://github.com/parcel-bundler/parcel/pull/7390
